### PR TITLE
fix: better scopes in unicorn, thread leakage in uniresis

### PR DESCRIPTION
**Pull request by antmat**
subj @3Hren @karitra @shaitan PTAL

### CODE REVIEW

File: unicorn/src/unicorn/zookeeper.cpp [hunk 1]
```diff
@@ -82,6 +80,21 @@ class scope_t: public api::unicorn_scope_t {
...
+public:
+    scope_wrapper_t(std::shared_ptr<scope_t> wrapped) :
+        wrapped(std::move(wrapped)) { }
+
+    ~scope_wrapper_t() {
+        wrapped->close();
+    }
+
+    auto close() -> void override {
+        wrapped->close();
```
Comment (3Hren) [inline unicorn/src/unicorn/zookeeper.cpp:94]: Double close. It is okay?

File: uniresis/src/uniresis.cpp [hunk 2]
```diff
@@ -89,14 +89,20 @@ class uniresis_t::updater_t : public std::enable_shared_from_this<uniresis_t::up
         subscope(),
         executor(),
         timer(executor.asio()),
-        log(std::move(log))
+        log(log)
     {}
 
+    ~updater_t() {
+        scope->close();
```
Comment (3Hren) [inline uniresis/src/uniresis.cpp]: Constructor doesn't initializes them, so SIGSEGVs nearby.

File: uniresis/src/uniresis.cpp [hunk 3]
```diff
@@ -89,14 +89,20 @@ class uniresis_t::updater_t : public std::enable_shared_from_this<uniresis_t::up
...
+        subscope->close();
+        timer.cancel();
+    }
+
     auto
     notify() -> void {
         COCAINE_LOG_DEBUG(log, "schedule resource notification on `{}` ...", path);
         scope = unicorn->create(
-            std::bind(&updater_t::on_create, shared_from_this(), ph::_1),
+            std::bind(&updater_t::on_create, this, ph::_1),
```
Comment (3Hren) [inline uniresis/src/uniresis.cpp:109]: No, you can't do this unless you own the I/O loop.

**Review by 3Hren — CHANGES_REQUESTED** [review]

File: uniresis/src/uniresis.cpp (hunk 3 above)
Comment (antmat) [reply uniresis/src/uniresis.cpp:109]: I own cancel token and it is thread safe - after close call no more callbacks will be dispatched and executing ones will end before close return.

File: uniresis/src/uniresis.cpp (hunk 2 above)
Comment (antmat) [reply uniresis/src/uniresis.cpp]: yep, here you are right, will fix tomorrow.

File: unicorn/src/unicorn/zookeeper.cpp (hunk 1 above)
Comment (antmat) [reply unicorn/src/unicorn/zookeeper.cpp:94]: it's ok

File: uniresis/src/uniresis.cpp (hunk 3 above)
Comment (3Hren) [reply uniresis/src/uniresis.cpp:109]: Imagine a situation when this callback is called while we're destructing ourself.

Comment (antmat) [reply uniresis/src/uniresis.cpp:109]: so what? destruction of members do not start until dtor body has been executed, and dtor cancelss callback.

**Pull request merged**